File: src/org/github/erikzielke/gotoproject/GotoProjectApplicationConfigurable.java
package org.github.erikzielke.gotoproject;

import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurationException;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

/**
 * Created by ez on 02/05/16.
 */
public class GotoProjectApplicationConfigurable  implements Configurable {
    private GotToProjectWindowSettingsForm settingsForm;


    @Nls
    @Override
    public String getDisplayName() {
        return "Go To Project Window";
    }

    @Nullable
    @Override
    public String getHelpTopic() {
        return null;
    }


    @Nullable
    @Override
    public JComponent createComponent() {
        if (settingsForm == null) {
            settingsForm = new GotToProjectWindowSettingsForm();
        }
        return settingsForm.getRootComponent();
    }

    @Override
    public boolean isModified() {
        GoToProjectWindowSettings state = GoToProjectApplicationComponent.getInstance().getState();
        return  settingsForm.isModified(state);
    }


    @Override
    public void apply() throws ConfigurationException {
        if (settingsForm != null) {
            GoToProjectWindowSettings state = GoToProjectApplicationComponent.getInstance().getState();

            settingsForm.getData(state);
        }
    }

    @Override
    public void reset() {
        if (settingsForm != null) {
            GoToProjectWindowSettings state = GoToProjectApplicationComponent.getInstance().getState();

            settingsForm.setData(state);
        }
    }

    @Override
    public void disposeUIResources() {
        settingsForm = null;
    }



}

File: src/main/kotlin/org/github/erikzielke/gotoproject/GotoProjectApplicationConfigurable.kt
package org.github.erikzielke.gotoproject

import com.intellij.openapi.options.Configurable
import com.intellij.openapi.options.ConfigurationException
import com.intellij.openapi.ui.DialogPanel
import com.intellij.ui.dsl.builder.bindSelected
import com.intellij.ui.dsl.builder.panel
import javax.swing.JComponent

class GotoProjectApplicationConfigurable : Configurable {
    private var settingsForm: DialogPanel? = null

    override fun getDisplayName(): String = "Go To Project Window"

    override fun getHelpTopic(): String? = null

    override fun createComponent(): JComponent? {
        val state = GoToProjectApplicationComponent.instance.state
        if (settingsForm == null) {
            settingsForm = createSettingsForm(state)
        }
        return settingsForm
    }

    private fun createSettingsForm(state: GoToProjectWindowSettings) = panel {
        row {
            checkBox("Include recently opened projects").bindSelected(state::isIncludeRecent)
        }
        row {
            checkBox("Panel in search everywhere").bindSelected(state::panelInSearchEverywhere)
        }
    }

    override fun isModified(): Boolean {
        return settingsForm?.isModified() ?: false
    }

    @Throws(ConfigurationException::class)
    override fun apply() {
        settingsForm?.apply()
    }

    override fun reset() {
        settingsForm?.reset()
    }

    override fun disposeUIResources() {
        settingsForm = null
    }

}
